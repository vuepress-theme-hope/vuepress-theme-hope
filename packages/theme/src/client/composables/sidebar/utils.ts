import { h } from "vue";
import { useRoute } from "vue-router";
import { useIconPrefix } from "../themeData";
import AutoLink from "../../components/AutoLink";
import { isActiveLink } from "../../utils";

import type { VNode } from "vue";
import type {
  ResolvedSidebarItem,
  ResolvedSidebarHeaderItem,
  ResolvedSidebarGroupItem,
} from "../../../shared";

export const renderIcon = (icon?: string): VNode | null =>
  icon
    ? h("i", {
        class: ["iconfont", `${useIconPrefix().value}${icon}`],
      })
    : null;

export const renderItem = (
  config: ResolvedSidebarItem,
  props: VNode["props"]
): VNode => {
  return config.link
    ? // if the item has link, render it as `<AutoLink>`
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      h(AutoLink, {
        ...props,
        config,
      })
    : // if the item only has text, render it as `<p>`
      h("p", props, [renderIcon(config.icon), config.text]);
};

export const renderGroupHeader = (
  { collapsable, icon, text }: ResolvedSidebarGroupItem,
  open: boolean
): (VNode | null)[] => [
  // icon
  renderIcon(icon),
  // title
  h("span", { class: "title" }, text),
  // arrow
  collapsable
    ? h("span", { class: ["arrow", open ? "down" : "right"] }, text)
    : null,
];

export const renderChildren = (
  children: ResolvedSidebarHeaderItem[]
): VNode | null => {
  const route = useRoute();
  if (!children) return null;

  return h(
    "ul",
    { class: "sidebar-sub-headers" },
    children.map((child) => {
      const active = isActiveLink(route, child.link);

      return h("li", { class: "sidebar-sub-header" }, [
        renderItem(child, {
          class: {
            "sidebar-link": true,
            heading: true,
            active,
          },
        }),
        renderChildren(child.children),
      ]);
    })
  );
};
