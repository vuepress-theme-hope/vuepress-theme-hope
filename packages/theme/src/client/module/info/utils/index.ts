export * from "./define";
export * from "./editlink";
