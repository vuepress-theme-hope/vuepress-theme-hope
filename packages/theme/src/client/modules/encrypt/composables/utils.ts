import { type ComputedRef, computed } from "vue";

import { useThemeData } from "@theme-hope/composables/index";

import { type EncryptConfig } from "../../../../shared/index.js";

export const useEncryptData = (): ComputedRef<EncryptConfig> => {
  const themeData = useThemeData();

  return computed(() => themeData.value.encrypt || {});
};
