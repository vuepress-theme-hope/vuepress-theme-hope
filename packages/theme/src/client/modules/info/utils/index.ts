export * from "./editLink.js";
export * from "./typings.js";
