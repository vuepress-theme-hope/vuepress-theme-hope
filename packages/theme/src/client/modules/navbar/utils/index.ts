import {
  type AutoLinkOptions,
  type NavGroup,
  type NavbarItem,
} from "../../../../shared/index.js";

// resolved
export type ResolvedThemeNavbarItem =
  | NavbarItem
  | NavGroup<AutoLinkOptions | NavGroup<AutoLinkOptions>>;
