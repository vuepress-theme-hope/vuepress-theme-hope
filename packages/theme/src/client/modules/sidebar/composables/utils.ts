import { ensureEndingSlash } from "@vuepress/shared";

export const resolvePrefix = (prefix = "", path = ""): string =>
  path.startsWith("/") ? path : `${ensureEndingSlash(prefix)}${path}`;
