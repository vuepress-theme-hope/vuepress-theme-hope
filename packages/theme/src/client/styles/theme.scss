@use "vuepress-shared/styles/wrapper";
@use "@sass-palette/hope-config";

@media (min-width: hope-config.$laptop) {
  .chart-wrapper,
  .flowchart-wrapper,
  .mermaid-wrapper {
    &::-webkit-scrollbar {
      width: 8px;
      height: 8px;
    }

    &::-webkit-scrollbar-track-piece {
      -webkit-border-radius: 8px;
      border-radius: 8px;
      background: rgb(0 0 0 / 10%);
    }
  }
}

// rtl
#{hope-config.$rtl-selector} a.header-anchor {
  float: right;
}

// docsearch fix
// stylelint-disable selector-class-pattern
#docsearch-container {
  min-width: 145.7px !important;

  @media (max-width: hope-config.$pad) {
    min-width: 36px !important;
  }
}

.DocSearch {
  &.DocSearch-Button {
    margin-left: 0;

    @media (max-width: hope-config.$pad) {
      min-width: 36px !important;
    }
  }

  .DocSearch-Button-Placeholder {
    display: inline-block;
    padding: 4px 12px 4px 6px;
    font-size: 14px;

    @media (max-width: 719px) {
      display: none;
    }
  }

  .DocSearch-Search-Icon {
    width: 1.25em;
    height: 1.25em;
  }

  .DocSearch-Button-Keys {
    @media (max-width: hope-config.$pad) {
      display: none;
    }
  }

  .DocSearch-Button-Key {
    background: var(--bg-color);
    box-shadow: none;
  }
}
// stylelint-enable selector-class-pattern
