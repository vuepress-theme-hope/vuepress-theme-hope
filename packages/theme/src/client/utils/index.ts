export * from "./getAncestorLinks.js";
export * from "./resolveLinkInfo.js";
