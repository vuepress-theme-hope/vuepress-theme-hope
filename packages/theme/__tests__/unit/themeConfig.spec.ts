/* eslint-disable @typescript-eslint/naming-convention */
import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, it, expect } from "vitest";

import { getStatus } from "../../src/node/status";
import { getThemeConfig } from "../../src/node/themeConfig";
import { emptyTheme } from "./__fixtures__/theme/empty";

import type { HopeThemeOptions } from "../../src/shared";

const app = createBaseApp({
  bundler: {} as any,
  source: path.resolve(__dirname, "./__fixtures__/src"),
  theme: emptyTheme,
});

describe("should generate themeConfig correcly", () => {
  it("Should contain basic properties", () => {
    const themeConfig = {};

    expect(getThemeConfig(app, themeConfig, getStatus({}))).toMatchSnapshot();
  });

  it("Should handle single language", () => {
    const themeConfig = {
      navbar: ["/", "/about"],
      sidebar: ["/", "/about"],
    };

    expect(getThemeConfig(app, themeConfig, getStatus({}))).toMatchSnapshot();
  });

  it("locale should have higher property", () => {
    const themeConfig = {
      navbar: ["/", "/about"],

      locales: {
        "/": {
          navbar: ["/", "/guide", "/about"],
        },
        "/zh/": {
          navbar: ["/zh/", "/zh/about"],
        },
      },
    };

    expect(getThemeConfig(app, themeConfig, getStatus({}))).toMatchSnapshot();
  });

  it("should fallback to root if locale config is missing", () => {
    const themeConfig = {
      navbar: ["/", "/about"],

      locales: {
        "/": {},
        "/zh/": {
          navbar: ["/zh/", "/zh/about"],
        },
      },
    };

    expect(getThemeConfig(app, themeConfig, getStatus({}))).toMatchSnapshot();
  });

  it("root only option should not appear in locales", () => {
    const themeConfig: HopeThemeOptions = {
      pure: true,
      darkmode: "disable",
      encrypt: {},
    };

    expect(getThemeConfig(app, themeConfig, getStatus({}))).toMatchSnapshot();
  });
});
