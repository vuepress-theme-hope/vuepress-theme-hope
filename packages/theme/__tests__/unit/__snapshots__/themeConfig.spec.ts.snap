// Vitest Snapshot v1

exports[`should generate themeConfig correcly > Should contain basic properties 1`] = `
{
  "blog": {},
  "darkmode": "switch",
  "encrypt": {},
  "fullscreen": false,
  "locales": {
    "/": {
      "blog": {},
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "placeholder": "Enter password",
        "remember": "Remember password",
        "title": "Page Encrypted",
      },
      "headerDepth": 2,
      "hideSiteNameonMobile": true,
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbarAutoHide": "mobile",
      "navbarIcon": true,
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
        "selectLangText": "Language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "repoDisplay": true,
      "routeLocales": {
        "404msg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "back": "Go back",
        "home": "Take me home",
        "openInNewWindow": "Open in new window",
      },
      "sidebar": "structure",
      "sidebarIcon": true,
    },
  },
  "pure": false,
  "themeColor": false,
}
`;

exports[`should generate themeConfig correcly > Should handle single language 1`] = `
{
  "blog": {},
  "darkmode": "switch",
  "encrypt": {},
  "fullscreen": false,
  "locales": {
    "/": {
      "blog": {},
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "placeholder": "Enter password",
        "remember": "Remember password",
        "title": "Page Encrypted",
      },
      "headerDepth": 2,
      "hideSiteNameonMobile": true,
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbar": [
        "/",
        "/about",
      ],
      "navbarAutoHide": "mobile",
      "navbarIcon": true,
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
        "selectLangText": "Language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "repoDisplay": true,
      "routeLocales": {
        "404msg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "back": "Go back",
        "home": "Take me home",
        "openInNewWindow": "Open in new window",
      },
      "sidebar": [
        "/",
        "/about",
      ],
      "sidebarIcon": true,
    },
  },
  "pure": false,
  "themeColor": false,
}
`;

exports[`should generate themeConfig correcly > locale should have higher property 1`] = `
{
  "blog": {},
  "darkmode": "switch",
  "encrypt": {},
  "fullscreen": false,
  "locales": {
    "/": {
      "blog": {},
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "placeholder": "Enter password",
        "remember": "Remember password",
        "title": "Page Encrypted",
      },
      "headerDepth": 2,
      "hideSiteNameonMobile": true,
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbar": [
        "/",
        "/guide",
        "/about",
      ],
      "navbarAutoHide": "mobile",
      "navbarIcon": true,
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
        "selectLangText": "Language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "repoDisplay": true,
      "routeLocales": {
        "404msg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "back": "Go back",
        "home": "Take me home",
        "openInNewWindow": "Open in new window",
      },
      "sidebar": "structure",
      "sidebarIcon": true,
    },
  },
  "pure": false,
  "themeColor": false,
}
`;

exports[`should generate themeConfig correcly > root only option should not appear in locales 1`] = `
{
  "blog": {},
  "darkmode": "disable",
  "encrypt": {},
  "fullscreen": false,
  "locales": {
    "/": {
      "blog": {},
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "placeholder": "Enter password",
        "remember": "Remember password",
        "title": "Page Encrypted",
      },
      "headerDepth": 2,
      "hideSiteNameonMobile": true,
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbarAutoHide": "mobile",
      "navbarIcon": true,
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
        "selectLangText": "Language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "repoDisplay": true,
      "routeLocales": {
        "404msg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "back": "Go back",
        "home": "Take me home",
        "openInNewWindow": "Open in new window",
      },
      "sidebar": "structure",
      "sidebarIcon": true,
    },
  },
  "pure": true,
  "themeColor": false,
}
`;

exports[`should generate themeConfig correcly > should fallback to root if locale config is missing 1`] = `
{
  "blog": {},
  "darkmode": "switch",
  "encrypt": {},
  "fullscreen": false,
  "locales": {
    "/": {
      "blog": {},
      "encryptLocales": {
        "errorHint": "Please enter the correct password!",
        "placeholder": "Enter password",
        "remember": "Remember password",
        "title": "Page Encrypted",
      },
      "headerDepth": 2,
      "hideSiteNameonMobile": true,
      "lang": "en-US",
      "metaLocales": {
        "author": "Author",
        "category": "Category",
        "contributors": "Contributors",
        "date": "Writing Date",
        "editLink": "Edit this page",
        "lastUpdated": "Last update",
        "next": "Next",
        "origin": "Original",
        "prev": "Prev",
        "readingTime": "Reading Time",
        "tag": "Tag",
        "toc": "On This Page",
        "views": "Page views",
        "words": "Words",
      },
      "navbar": [
        "/",
        "/about",
      ],
      "navbarAutoHide": "mobile",
      "navbarIcon": true,
      "navbarLocales": {
        "langName": "English",
        "selectLangAriaLabel": "Select language",
        "selectLangText": "Language",
      },
      "outlookLocales": {
        "darkmode": "Theme Mode",
        "fullscreen": "Full Screen",
        "themeColor": "Theme Color",
      },
      "repoDisplay": true,
      "routeLocales": {
        "404msg": [
          "There’s nothing here.",
          "How did we get here?",
          "That’s a Four-Oh-Four.",
          "Looks like we've got some broken links.",
        ],
        "back": "Go back",
        "home": "Take me home",
        "openInNewWindow": "Open in new window",
      },
      "sidebar": "structure",
      "sidebarIcon": true,
    },
  },
  "pure": false,
  "themeColor": false,
}
`;
